p2p/pubsub/psHost.go

```diff
@@ -0,0 +1,366 @@
+	"github.com/rs/zerolog"
+)
+
+// psHost tasks. Each task represents a upper function call which is executed in a single
+// thread.
+type (
+	addHandlerTask struct {
+		handler Handler
+		errC    chan error
+	}
+	startHandlerTask struct {
+		spec HandlerSpecifier
+		errC chan error
+	}
+	stopHandlerTask struct {
+		spec HandlerSpecifier
+		errC chan error
+	}
+	removeHandlerTask struct {
+		spec HandlerSpecifier
+		errC chan error
+	}
+	removeTopicTask struct {
+		topic Topic
+		errC  chan error
+	}
+	closeTask struct {
+		closed chan struct{}
+	}
+)
+
+// pubSubHost is the host used for pub-sub message handling.
+// All requests in pubSubHost are handled single threaded.
+type pubSubHost struct {
+	pubSub         rawPubSub
+	optionProvider ValidateOptionProvider
+
+	// registered topic and handlers
+	topicRunners map[Topic]*topicRunner       // all topicRunners (running & stopped)
+	handlers     map[HandlerSpecifier]Handler // all pubSubHandlers (running & stopped)
+
+	// task channels
+	addHandlerC    chan addHandlerTask
+	startHandlerC  chan startHandlerTask
+	stopHandlerC   chan stopHandlerTask
+	removeHandlerC chan removeHandlerTask
+	removeTopicC   chan removeTopicTask
+	closeTaskC     chan closeTask
+
+	// utils
+	topicLock sync.RWMutex // handlerLock to protect topicRunners
+	log       zerolog.Logger
+}
+
+func newPubSubHost(pubSub *libp2p_pubsub.PubSub, optionProvider ValidateOptionProvider, log zerolog.Logger) *pubSubHost {
+	return &pubSubHost{
+		pubSub:         newPubSubAdapter(pubSub),
+		optionProvider: optionProvider,
+
+		topicRunners: make(map[Topic]*topicRunner),
+		handlers:     make(map[HandlerSpecifier]Handler),
+
+		addHandlerC:    make(chan addHandlerTask),
+		startHandlerC:  make(chan startHandlerTask),
+		stopHandlerC:   make(chan stopHandlerTask),
+		removeHandlerC: make(chan removeHandlerTask),
+		removeTopicC:   make(chan removeTopicTask),
+		closeTaskC:     make(chan closeTask),
+
+		log: log.With().Str("module", "pub-sub").Logger(),
+	}
+}
+
+// Start start the pubSubHost to handler requests (a.k.a. upper function calls).
+// All upper function calls are running in a single thread to ensure the race
+// condition is not a concern.
+func (psh *pubSubHost) Start() {
+	go psh.run()
+}
+
+func (psh *pubSubHost) run() {
+	for {
+		select {
+		case task := <-psh.addHandlerC:
+			task.errC <- psh.addHandler(task.handler)
+
+		case task := <-psh.startHandlerC:
+			task.errC <- psh.startHandler(task.spec)
+
+		case task := <-psh.stopHandlerC:
+			task.errC <- psh.stopHandler(task.spec)
+
+		case task := <-psh.removeHandlerC:
+			task.errC <- psh.removeHandler(task.spec)
+
+		case task := <-psh.removeTopicC:
+			task.errC <- psh.removeTopic(task.topic)
+
+		case task := <-psh.closeTaskC:
+			close(task.closed)
+			return
+		}
+	}
+}
+
+// Close close the pubSubHost.
+func (psh *pubSubHost) Close() {
+	task := closeTask{closed: make(chan struct{})}
+	psh.closeTaskC <- task
+
+	<-task.closed
```

Comment:
why is this needed?

Reply:
This is to wait for the close task to be executed at `run`.